Add Support for Sending Basic Native Collections

DavidMohrhardt opened the pull request:
This adds support to the `RTCDataChannel` for sending the basic Native Collections `NativeArray<T>`, `NativeSlice<T>` and `NativeArray<T>.ReadOnly`.

- Adds `Send` functionality to `RTCDataChannel` for each native collection described above.
- Adds `Send` functionality to `RTCDataChannel` for arbitrary pointers in the `unsafe` context and `IntPtr`s.
- Adds Native Collection send test cases where applicable.

Currently is in a first pass and not completely ready for review but tests are passing.

karasusan commented on Plugin~/WebRTCPlugin/WebRTCPlugin.cpp:
```diff
@@ -1346,6 +1346,11 @@ extern "C"
         dataChannelObj->Send(msg);
     }
 
+    UNITY_INTERFACE_EXPORT void DataChannelSendPtr(DataChannelObject* dataChannelObj, const byte* msg, int len)
```
It looks like `DataChannelSendBinary` function same.

DavidMohrhardt replied:
It is, I forgot about [DllImportAttribute.EntryPoint](https://docs.microsoft.com/en-us/dotnet/api/system.runtime.interopservices.dllimportattribute.entrypoint?view=net-5.0).  Fixed in next commit.

karasusan commented:
I checked the result of testing on CI and This error is not related your pull request.

I am sure why the testcase `Unity.WebRTC.RuntimeTest.TransceiverTest.TransceiverStop` is failed, but we will fix it soon.

karasusan commented:
@DavidMohrhardt
I noticed this error might be related the change.

https://unity-ci.cds.internal.unity3d.com/job/9956621

```
Message: Expected: True But was: False StackTrace: at Unity.WebRTC.RuntimeTest.DataChannelTest+<SendAndReceiveMessageWithExecuteTasks>d__11.MoveNext () [0x00000] in <00000000000000000000000000000000>:0 at UnityEngine.TestTools.TestEnumerator+<Execute>d__7.MoveNext () [0x00000] in <00000000000000000000000000000000>:0 at UnityEngine.TestTools.EnumerableTestMethodCommand+<ExecuteEnumerableAndRecordExceptions>d__4.MoveNext () [0x00000] in <00000000000000000000000000000000>:0 at UnityEngine.TestTools.EnumerableTestMethodCommand.AdvanceEnumerator (System.Collections.IEnumerator enumerator) [0x00000] in <00000000000000000000000000000000>:0 at UnityEngine.TestTools.EnumerableTestMethodCommand+<ExecuteEnumerable>d__2.MoveNext () [0x00000] in <00000000000000000000000000000000>:0 at UnityEngine.TestRunner.NUnitExtensions.Runner.UnityLogCheckDelegatingCommand+<ExecuteEnumerable>d__3.MoveNext () [0x00000] in <00000000000000000000000000000000>:0 at UnityEngine.TestTools.BeforeAfterTestCommandBase`1+<ExecuteEnumerable>d__15[T].MoveNext () [0x00000] in <00000000000000000000000000000000>:0 at UnityEngine.TestTools.BeforeAfterTestCommandBase`1+<ExecuteEnumerable>d__15[T].MoveNext () [0x00000] in <00000000000000000000000000000000>:0 at UnityEngine.TestTools.BeforeAfterTestCommandBase`1+<ExecuteEnumerable>d__15[T].MoveNext () [0x00000] in <00000000000000000000000000000000>:0 at UnityEngine.TestTools.BeforeAfterTestCommandBase`1+<ExecuteEnumerable>d__15[T].MoveNext () [0x00000] in <00000000000000000000000000000000>:0 at UnityEngine.SetupCoroutine.InvokeMoveNext (System.Collections.IEnumerator enumerator, System.IntPtr returnValueAddress) [0x00000] in <00000000000000000000000000000000>:0 Test Run Information: * Run count: 3 * Failed count: 1 * Success count: 2
```

karasusan commented:
related issue
https://github.com/Unity-Technologies/com.unity.webrtc/issues/308

karasusan commented:
memo: WRS-23

DavidMohrhardt commented:
> @DavidMohrhardt I noticed this error might be related the change.
> 
> https://unity-ci.cds.internal.unity3d.com/job/9956621
> 
> ```
> Message: Expected: True But was: False StackTrace: at Unity.WebRTC.RuntimeTest.DataChannelTest+<SendAndReceiveMessageWithExecuteTasks>d__11.MoveNext () [0x00000] in <00000000000000000000000000000000>:0 at UnityEngine.TestTools.TestEnumerator+<Execute>d__7.MoveNext () [0x00000] in <00000000000000000000000000000000>:0 at UnityEngine.TestTools.EnumerableTestMethodCommand+<ExecuteEnumerableAndRecordExceptions>d__4.MoveNext () [0x00000] in <00000000000000000000000000000000>:0 at UnityEngine.TestTools.EnumerableTestMethodCommand.AdvanceEnumerator (System.Collections.IEnumerator enumerator) [0x00000] in <00000000000000000000000000000000>:0 at UnityEngine.TestTools.EnumerableTestMethodCommand+<ExecuteEnumerable>d__2.MoveNext () [0x00000] in <00000000000000000000000000000000>:0 at UnityEngine.TestRunner.NUnitExtensions.Runner.UnityLogCheckDelegatingCommand+<ExecuteEnumerable>d__3.MoveNext () [0x00000] in <00000000000000000000000000000000>:0 at UnityEngine.TestTools.BeforeAfterTestCommandBase`1+<ExecuteEnumerable>d__15[T].MoveNext () [0x00000] in <00000000000000000000000000000000>:0 at UnityEngine.TestTools.BeforeAfterTestCommandBase`1+<ExecuteEnumerable>d__15[T].MoveNext () [0x00000] in <00000000000000000000000000000000>:0 at UnityEngine.TestTools.BeforeAfterTestCommandBase`1+<ExecuteEnumerable>d__15[T].MoveNext () [0x00000] in <00000000000000000000000000000000>:0 at UnityEngine.TestTools.BeforeAfterTestCommandBase`1+<ExecuteEnumerable>d__15[T].MoveNext () [0x00000] in <00000000000000000000000000000000>:0 at UnityEngine.SetupCoroutine.InvokeMoveNext (System.Collections.IEnumerator enumerator, System.IntPtr returnValueAddress) [0x00000] in <00000000000000000000000000000000>:0 Test Run Information: * Run count: 3 * Failed count: 1 * Success count: 2
> ```

So since this seems like an instability I am gonna rework the tests to better tell which portion of the test is failing.  I haven't been able to repro it locally but I will try again both on Yamato and locally.

DavidMohrhardt commented:
> > @DavidMohrhardt I noticed this error might be related the change.
> > https://unity-ci.cds.internal.unity3d.com/job/9956621
> > ```
> > Message: Expected: True But was: False StackTrace: at Unity.WebRTC.RuntimeTest.DataChannelTest+<SendAndReceiveMessageWithExecuteTasks>d__11.MoveNext () [0x00000] in <00000000000000000000000000000000>:0 at UnityEngine.TestTools.TestEnumerator+<Execute>d__7.MoveNext () [0x00000] in <00000000000000000000000000000000>:0 at UnityEngine.TestTools.EnumerableTestMethodCommand+<ExecuteEnumerableAndRecordExceptions>d__4.MoveNext () [0x00000] in <00000000000000000000000000000000>:0 at UnityEngine.TestTools.EnumerableTestMethodCommand.AdvanceEnumerator (System.Collections.IEnumerator enumerator) [0x00000] in <00000000000000000000000000000000>:0 at UnityEngine.TestTools.EnumerableTestMethodCommand+<ExecuteEnumerable>d__2.MoveNext () [0x00000] in <00000000000000000000000000000000>:0 at UnityEngine.TestRunner.NUnitExtensions.Runner.UnityLogCheckDelegatingCommand+<ExecuteEnumerable>d__3.MoveNext () [0x00000] in <00000000000000000000000000000000>:0 at UnityEngine.TestTools.BeforeAfterTestCommandBase`1+<ExecuteEnumerable>d__15[T].MoveNext () [0x00000] in <00000000000000000000000000000000>:0 at UnityEngine.TestTools.BeforeAfterTestCommandBase`1+<ExecuteEnumerable>d__15[T].MoveNext () [0x00000] in <00000000000000000000000000000000>:0 at UnityEngine.TestTools.BeforeAfterTestCommandBase`1+<ExecuteEnumerable>d__15[T].MoveNext () [0x00000] in <00000000000000000000000000000000>:0 at UnityEngine.TestTools.BeforeAfterTestCommandBase`1+<ExecuteEnumerable>d__15[T].MoveNext () [0x00000] in <00000000000000000000000000000000>:0 at UnityEngine.SetupCoroutine.InvokeMoveNext (System.Collections.IEnumerator enumerator, System.IntPtr returnValueAddress) [0x00000] in <00000000000000000000000000000000>:0 Test Run Information: * Run count: 3 * Failed count: 1 * Success count: 2
> > ```
> 
> So since this seems like an instability I am gonna rework the tests to better tell which portion of the test is failing. I haven't been able to repro it locally but I will try again both on Yamato and locally.

@karasusan turns out, this is a non-issue based around what I think was a misunderstanding. `WebRTC.ExecutePendingTasks()` will return false if there are still tasks that need doing but the inner workings of that are that whenever the `ExecutableSynchronizationContext` executes work, it takes the entire queue of work items when the method is called, processes all of them, then checks if it still has time left to do another batch. If another thread creates more work and puts it in the queue during the first batch then the result of the method will be false. This is inline with the behavior of the synchronization context that Unity uses.

So the tests in question should likely be using a `StopWatch` timeout and checking if the messages were received rather than checking if there might be potentially unrelated work still in the synchronization context to do. I updated them to do so but I want to double check with you if that's the behavior you want.

karasusan commented:
I found the compile error on Unity2019.
I think this error is contained commits in this PR.
https://github.com/Unity-Technologies/com.unity.webrtc/pull/522

![image](https://user-images.githubusercontent.com/1132081/142363673-02189c15-c0fc-4a15-a013-e796b3d621c5.png)

There is the trigger for testing Unity2019, but the trigger isn't kicked automatically for saving machine resources in CI.

DavidMohrhardt commented:
Those correspond to the following lines where I was using null-coalescing and using declarations:

https://github.com/Unity-Technologies/com.unity.webrtc/blob/586a9fc25c4f6fd3088465457340fe108d369e4d/Runtime/Scripts/Internal/ExecutableUnitySynchronizationContext.cs#L30
https://github.com/Unity-Technologies/com.unity.webrtc/blob/586a9fc25c4f6fd3088465457340fe108d369e4d/Runtime/Scripts/Internal/ExecutableUnitySynchronizationContext.cs#L67

I had forgotten that 19.4 doesn't support these 2 things.  I also updated the native collections sending API to guard against `NativeArray<T>.ReadOnly` since that was also introduced in 20.1.  Great catch!
